web/source/text/processor.ts

```diff
@@ -30,49 +30,84 @@ namespace com.keyman.text {
       }
 
-      // If this was triggered by the OSK -or- if it was triggered by a 'synthetic' OutputTarget (TouchAlias, Mock)
-      // that lacks default key processing behavior.
+      let matched = false;
+      var char = '';
       if(usingOSK || outputTarget.isSynthetic) {
-        let result = DefaultOutput.applyCommand(Lkc, keyShiftState);
-
-        if(result) {
-          return result;
+        matched = true;  // All the conditions below result in matches until the final else, which restores the expected default
+                         // if no match occurs.
+
+        if(DefaultOutput.isCommand(Lkc)) {
+          // Note this in the rule behavior, return successfully.  We'll consider applying it later.
+          ruleBehavior.triggersDefaultCommand = true;
+
+          // We'd rather let the browser handle these keys, but we're using emulated keystrokes, forcing KMW
+          // to emulate default behavior here.
+        } else if(char = DefaultOutput.forSpecialEmulation(Lkc)) { 
+          switch(char) {
+            case '\b':
+              keyman.interface.defaultBackspace(outputTarget);
+              break;
+            case '\n':
+              outputTarget.handleNewlineAtCaret();
+              break;
+            case ' ':
+              keyman.interface.output(0, outputTarget, ' ');
+              break;
+            // case '\u007f': // K_DEL
+              // // For (possible) future implementation.
+              // // Would recommend (conceptually) equaling K_RIGHT + K_BKSP, the former of which would technically be a 'command'.
+            default:
+              ruleBehavior.errorLog = "Unexpected 'special emulation' character (\\u" + char.kmwCharCodeAt(0).toString(16) + ") went unhandled!";
+          } 
+        } else {
+          // Back to the standard default, pending normal matching.
+          matched = false;
         }
-      } else if(Lkc.Lcode == 8) {
-        keyman.interface.defaultBackspace();
-        return '';
       }
 
-      let code: string;
+      if(!matched) {
+        if(char = DefaultOutput.forAny(Lkc, keyShiftState)) {
+          if(char == '\b') { // physical keystrokes.
+            keyman.interface.defaultBackspace(outputTarget);
+          } else {
+            keyman.interface.output(0, outputTarget, char);
+          }
+        } else {
+          // No match, no default RuleBehavior.
+          return null;
+        }
+      }
 
-      // Translate numpad keystrokes into their non-numpad equivalents
-      if(code = DefaultOutput.forNumpadKeys(Lkc)) {
-        return code;
-      } else if(code = DefaultOutput.forUnicodeKeynames(Lkc)) { // Test for fall back to U_xxxxxx key id
-        return code;
-      } else if(code = DefaultOutput.forBaseKeys(Lkc, keyShiftState)) {
-        return code;
+      // Shortcut things immediately if there were issues generating this rule behavior.
+      if(ruleBehavior.errorLog) {
```

Comment:
In this scenario, we effectively get a 'null' ruleBehavior right? Is it better to give a last-ditch default behavior instead?

Reply:
I'll have to double-check, but I'm fairly certain that this _is_ the current "last-ditch default behavior" I saw when I traced through the logic.  (No actual text produced whenever warning/error logs produced.)

Reply:
OK.